node/core/provisioner/src/disputes/prioritized_selection/mod.rs

```diff
@@ -0,0 +1,460 @@
+		};
+	}
+
+	// ... and INACTIVE in three more
+	for (session_index, candidate_hash, _) in inactive {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(onchain_state) =>
+				if concluded_onchain(onchain_state) {
+					partitioned.inactive_concluded_onchain.push((session_index, candidate_hash));
+				} else {
+					partitioned.inactive_unconcluded_onchain.push((session_index, candidate_hash));
+				},
+			None => partitioned.inactive_unknown_onchain.push((session_index, candidate_hash)),
+		}
+	}
+
+	partitioned
+}
+
+// Helper trait to obtain the value of vote for `InvalidDisputeStatementKind` and `ValidDisputeStatementKind`.
+// The alternative was to pass a bool to `fn is_vote_worth_to_keep` explicitly but it's pointless as the value is already 'encoded' in the type.
+trait VoteType {
+	fn is_valid() -> bool;
+}
+
+impl VoteType for InvalidDisputeStatementKind {
+	fn is_valid() -> bool {
+		false
+	}
+}
+
+impl VoteType for ValidDisputeStatementKind {
+	fn is_valid() -> bool {
+		true
+	}
+}
+
+/// Determines if a vote is worth to be kept, based on the onchain disputes
+fn is_vote_worth_to_keep<T: VoteType>(
+	validator_index: &ValidatorIndex,
+	_: &T,
+	onchain_state: &DisputeState,
+) -> bool {
+	let offchain_vote = T::is_valid();
+	let in_validators_for = onchain_state
+		.validators_for
+		.get(validator_index.0 as usize)
+		.as_deref()
+		.copied()
+		.unwrap_or(false);
+	let in_validators_against = onchain_state
+		.validators_against
+		.get(validator_index.0 as usize)
+		.as_deref()
+		.copied()
+		.unwrap_or(false);
+
+	if in_validators_for && in_validators_against {
+		// The validator has double voted and runtime knows about this. Ignore this vote.
+		return false
+	}
+
+	if offchain_vote && in_validators_against || !offchain_vote && in_validators_for {
+		// offchain vote differs from the onchain vote
+		// we need this vote to punish the offending validator
+		return true
+	}
+
+	// The vote is valid. Return true if it is not seen onchain.
+	!in_validators_for && !in_validators_against
+}
+
+/// Request disputes identified by `CandidateHash` and the `SessionIndex`.
+async fn request_disputes(
+	sender: &mut impl overseer::ProvisionerSenderTrait,
+) -> Vec<(SessionIndex, CandidateHash, DisputeStatus)> {
+	let (tx, rx) = oneshot::channel();
+	let msg = DisputeCoordinatorMessage::RecentDisputes(tx);
+
+	// Bounded by block production - `ProvisionerMessage::RequestInherentData`.
+	sender.send_unbounded_message(msg);
+
+	let recent_disputes = rx.await.unwrap_or_else(|err| {
+		gum::warn!(target: LOG_TARGET, err=?err, "Unable to gather recent disputes");
+		Vec::new()
+	});
+	recent_disputes
+}
+
+// This function produces the return value for `pub fn select_disputes()`
+fn into_multi_dispute_statement_set(
+	metrics: &metrics::Metrics,
+	dispute_candidate_votes: BTreeMap<(SessionIndex, CandidateHash), CandidateVotes>,
+) -> Result<MultiDisputeStatementSet, Error> {
```

Comment:
No reason to return a Result, at least I don't see any errors being possible here.